### DISCUSSION TITLE
Relative auto-imports

### DISCUSSION THREAD

**Discussion by jaycle** [python-rope/rope#682] [Q&A] [answered]
Hi, I'm using rope through python-lsp-server and neovim.

My question is if it is possible to control how the auto import functionality builds the import statement? My `:LspInfo` puts my root directory at the top of my project. But I have separate python projects (and other languages) in directories beneath.

Auto-import works great, but it builds a path like `from server.app.models import MyModel` when within, say, `server/app/views.py` (typical Django project structure).

If this is a question for python-lsp-server, I'm happy to ask over there. Thanks!

**Comment by lieryan** [comment]
The current implementation of autoimport supports only absolute import. It should be possible to implement a setting to do relative import though this is not currently implemented yet, here is the place where [rope generate the import statement](https://github.com/python-rope/rope/blob/7e1c8971a42bfe0414f914b456ce242e2bea10a6/rope/contrib/autoimport/sqlite.py#L200-L204). 

For relative import, python-lsp-server would also need to pass in the filename of the importing module, so that would require a change in both rope and python-lsp-server.

cc: @bagel897

**Reply by bagel897** [reply]
I am personally against relative imports, especially for autoimport. For autoimport there's additional difficulty in generating these suggestions. Besides the argument passing, you'd need to find imports relative to each other with the filenames.  I did not design the parsing or utils modules to handle relative imports. Currently, they do not put filenames into the database. Part of this is that some modules are imported (i.e.: They used the compiled path) and probably cannot be used for relative imports.  Furthermore, it'd be confusing for the user to have two suggestions per completion.

If someone wanted to implement this, they'd be welcome to, the relevant parts are both _search_name and _search_module. I'd hide it behind a config flag which changes the default for completions in the project from absolute to relative. 

This is a rope issue, it'd first have to be implemented here.

**Reply by jaycle** [reply]
Thank you both for the helpful replies. While relative imports are useful to me, personally, I probably mistitled my question. I think I'm looking for the equivalent of what PyCharm calls "source root." ([ref](https://www.jetbrains.com/help/pycharm/content-root.html))

Basically, I'm hoping to skip that extra `server` from the import mentioned in the original question. I'm guessing the answer is still the same though as far as it being possible? But I'm curious if it influences how @bagel897 feels about it philosophically. The root of each "project" is just a bit lower in the directory tree than where the LSP is counting as its root.

**Reply by lieryan** [reply]
So, IIUC, you're actually just trying to set the rope's project root to be different than your Git/VCS project's root. That should be fairly straightforward then, you just need to set the project root as appropriate. How to do this is editor-specific, in neovim's built in LSP, you can [set the root_dir](https://neovim.io/doc/user/lsp.html#lsp-quickstart).

This seems more like an neovim lsp question on how to configure root_dir, rather than pylsp/rope in particular.

**Reply by jaycle** [reply]
Yes, I think that's a decent summary. Though, I had wondered/hoped that I could somehow control the "import root" without affecting the LSP root directory. Ideally, I wouldn't run a separate LSP per project since they do all, in fact, share the same `.venv` and `pyproject.toml`. 

Since rope has `source_folders` and `python_path` as config options, I didn't know if that could somehow influence the upper limit of the import. Perhaps our project is a little non-standard. But as far as I can tell, all other tooling works for us when we set `PYTHONPATH=server` or `PYTHONPATH=worker` where worker and server are subdirectories.

Actually, I typed the above and re-checked, and it seems that if I set `prefs["python_path"] = ["server", "~/projects/myco/.venv/lib/python3.11/site-packages"]` I get the import statement I wanted.

I really appreciate the responses and your opening of the ticket. I still see value in support of relative imports for rope/pylsp, but being able to control the "root" is what I was after so I'll mark this as answered.

**Reply by jaycle** [reply]
@lieryan Marked my own answer, just because I wanted to leave a summary for future readers. Please let me know if that misleads other users of the library. I'm happy to discuss further, but don't wish to take any more of your time. Thanks again for all the help!

**Reply by lieryan** [reply]
@jaycle thank you for writing how you solved the issue! That's really great.

**Comment by jaycle** [comment, marked as answer]
By setting the `python_path` [config option](https://rope.readthedocs.io/en/latest/configuration.html) to include the subfolder of your project, you can cause the auto-import to stop at the root provided.

### Example
- server/common/views.py
- server/main/api.py

Within api.py, importing from views.py would normally be `from server.common.views import MyView`. By setting `python_path` to include `server`, the import becomes  `from common.views import MyView`.
